injector/src/test/java/edu/ucr/cs/riple/injector/ComprehensiveTest.java

```diff
@@ -0,0 +1,294 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package edu.ucr.cs.riple.injector;
+
+import static java.util.Collections.singleton;
+
+import edu.ucr.cs.riple.injector.changes.AddMarkerAnnotation;
+import edu.ucr.cs.riple.injector.changes.RemoveAnnotation;
+import edu.ucr.cs.riple.injector.location.OnField;
+import edu.ucr.cs.riple.injector.location.OnMethod;
+import edu.ucr.cs.riple.injector.location.OnParameter;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class ComprehensiveTest extends BaseInjectorTest {
+
+  @Test
+  public void addFieldMethodParamTest() {
+    String annot = "javax.annotation.Nullable";
+    injectorTestHelper
+        .addInput(
+            "Test.java",
+            "package edu.ucr;",
+            "public class Test {",
+            "   Field foo1;",
+            "   Field foo2;",
+            "   Field foo3, foo4;",
+            "   Object bar1(Object p1, Object p2) {",
+            "       return new Object();",
+            "   }",
+            "   Object bar2(Object p1, Object p2) {",
+            "       return new Object();",
+            "   }",
+            "   Field foo5;",
+            "   Object bar3(Object p1, Object p2, Object p3, Object p4) {",
+            "       return new Object();",
+            "   }",
+            "   Object bar4() {",
+            "       return new Object();",
+            "   }",
+            "}")
+        .expectOutput(
+            "package edu.ucr;",
+            "import javax.annotation.Nullable;",
+            "public class Test {",
+            "   Field foo1;",
+            "   @Nullable",
+            "   Field foo2;",
+            "   @Nullable",
+            "   Field foo3, foo4;",
+            "   @Nullable",
+            "   Object bar1(@Nullable Object p1, Object p2) {",
+            "       return new Object();",
+            "   }",
+            "   Object bar2(Object p1, @Nullable Object p2) {",
+            "       return new Object();",
+            "   }",
+            "   @Nullable",
+            "   Field foo5;",
+            "   @Nullable",
+            "   Object bar3(@Nullable Object p1, @Nullable Object p2, Object p3, @Nullable Object p4) {",
+            "       return new Object();",
+            "   }",
+            "   @Nullable",
+            "   Object bar4() {",
+            "       return new Object();",
+            "   }",
+            "}")
```

Comment:
I was not able to fully reproduce the example you gave me, Would you please tell me how I can pass an interface as a parameter for annotation, it seems to it is not possible. 

Instead I used the example below:
```java
@Foo(clazz = String.class, value = "Some description"),
private static @Baz Object foo6(
   final @Baz Map<String, @Baz Object> m,
   @Foo(clazz = String.class, value = "Some argument") @Baz Object o1,
   Object o2,
   Object o3,
   @Foo(clazz = Object.class, value = \"Some other argument\") @Baz Object o4) {
 return new @Baz Runnable() {
   public void run() {
     System.out.print(\"log\");
   }
 };
}
```

Added 5 unit tests on method and parameters and all tests pass. 509aaae3d17204c92a4573aec5094b41669777dd

Reply:
> Would you please tell me how I can pass an interface as a parameter for annotation, it seems to it is not possible.

Not sure I understand the question. I think all you need is an annotation with `value` of type `Object` or `Runnable` to pass an anonymous subclass of `Runnable` to it. But, also, that probably isn't the most interesting part of my convoluted example, so it's fine to leave it out. I would, however, make sure `o1` and `o2` are on the same line, and similarly for `o3` and `o4`, to make sure we are testing the "column" position info in addition to the "line" position info. So:

```
@Foo(clazz = String.class, value = "Some description"),
private static @Baz Object foo6(
   final @Baz Map<String, @Baz Object> m,
   @Foo(clazz = String.class, value = "Some argument") @Baz Object o1, Object o2,
   Object o3, @Foo(clazz = Object.class, value = \"Some other argument\") @Baz Object o4) {
 return new @Baz Runnable() {
   public void run() {
     System.out.print(\"log\");
   }
 };
}
```

(While weird - and un-GJF'ed - formatting, that should test the injector a bit more)

Reply:
Interesting. You are right, can't be an arbitrary anonymous class, but it can be e.g. "an annotation type". Wonder if an anonymous subtype of an enum or an annotation is a thing... but either way, the example should be fine as is 🙂